KEP-2535: update the KEP with on-by-default strategy (#4789)
kubernetes/enhancements: keps/sig-node/2535-ensure-secret-pulled-images/README.md

- For `IfNotPresent` images, the kubelet will re-pull the image
- For `Never` images, the image creation will fail

This behavior mirrors what would happen if an image was not present on the node (as opposed to present, but yet to be authorized with the credentials present).

This will be enforced for both `IfNotPresent` and `Never` policies with the exception for images that were
pre-loaded on the node and the kubelet does not track them. An administrator will be allowed to modify this behavior
so that only a specified subset of these pre-loaded images can be pulled without re-authenticating.
**TODO**: describe the kubelet configuration option for allow-listing images

The goal is to enable this new behavior by default with an option to revert to the old one by using a kubelet configuration

> field `noPullImageSecretRecheck`.

Author: I restructured the section a bit but mostly tried to stick to your phrasing.

Optimize to only force re-authentication for a pod container image when the
`ImagePullSecrets` used to pull the container image has not already been authenticated.
IOW if an image is pulled with authentication for a first pod, subsequent pods that have the same
authentication information should not need to re-authenticate.

Images already present at boot or loaded externally to the kubelet will not require
authentication unless configured otherwise.
Images successfully pulled through the kubelet with no `ImagePullSecrets`/authentication required will
not require authentication.

The new behavior is designed in a way so that it replaces current behavior - it is going

> to be an on-by-default feature.

Reviewer: nit: clarify that it will be on-by-default after it graduates, this isn't proposing direct-to-beta-and-enabled

```suggestion
to be an on-by-default feature once graduated.
```

Optimize to only force re-authentication for a pod container image when the
`ImagePullSecrets` used to pull the container image has not already been authenticated.
IOW if an image is pulled with authentication for a first pod, subsequent pods that have the same
authentication information should not need to re-authenticate.

Images already present at boot or loaded externally to the kubelet will not require
authentication unless configured otherwise.

> Images successfully pulled through the kubelet with no `ImagePullSecrets`/authentication required will

Reviewer: ```suggestion
Images successfully pulled through the kubelet with no `ImagePullSecrets` authentication required will
```

This feature will not change the behavior of pod with `ImagePullPolicy` `Always`.

Enforcing periodical repull might be important in order to, for example, be able
to recheck image license entitlement, but is a non-goal for this enhancement.

## Proposal

> A new API will be created to represent metadata about image pulls being authenticated. This API

Reviewer: This makes it sound like there will be REST API surface, which isn't what we're proposing.

I think it's helpful to start by describing what the kubelet will do:

> The kubelet will track container images that required authentication and a list of the
authentications that successfully pulled the image. This list will persist across reboots of host and restart of
kubelet.

And then go into *how* it will do it

> This will be done using a files persisted within the kubelet directory on the node.
> The files will be structured and versioned using standard config file API versioning techniques.

will be used by the kubelet to persist this metadata in file(s) on the kubelet's node.
The API is a representation of the kubelet's in-memory state, where the kubelet tracks
information about image pulls being authenticated.

The kubelet will ensure any image it keeps a track of is always pulled if authentication information
from an image pull is not yet present, thus enforcing authentication / re-authentication.

A new kubelet configuration option will be added, as well as a feature
gate to gate its use:

> - `noPullImageSecretRecheck`

Reviewer: let's noodle a little bit on whether we can come up with a config knob / knobs that makes the spectrum of enforcement coherent:

How would each of these use cases be expressed in the kubelet config?
```
1. Do not allow access to _any_ pre-loaded images without verification
2. Only allow access to a _specific_ list of pre-loaded images without verification
3. Allow access to _all_ pre-loaded and pulled-without-pull-secrets images without verification (default behavior when this feature is enabled)
4. Allow access to all images _without verification_ (default behavior when this feature is not enabled, not recommended)
```

(again, I'm not 100% sure we want to make number 4 possible)

strawman possibilities (I didn't spend a ton of time on this, just wanted to start you thinking):

For 1:
```yaml
preexistingImagePolicy: VerifyAccess
```

For 2:
```yaml
preexistingImagePolicy: VerifyAccessUnlessAllowlisted
preexistingImageAllowlist:
- myregistry.example.com/myimage:mytag
- myregistry.example.com/myotherimage:*
- mypublicregistry.example.com/*
```

For 3 (default policy):
```yaml
preexistingImagePolicy: AllowAccessWithoutVerification
```

Author: Given the scenarios we've got, that's actually a good option list. I think I'd probably only name it differently, maybe structure it a bit more for simpler config validation:
```yaml
disableImageSecretPullRecheck: true | false # in case we actually want to be able to disable this
preexistingImagesPullPolicy:
    type: AlwaysVerify | IgnoreListed | NeverVerify
    ignoreList: # specific to image tag / just an image + all tags / whole registry
    - myregistry.example.com/myimage:mytag
    - myregistry.example.com/myotherimage:*
    - mypublicregistry.example.com/*
```
```go
type KubeletConfiguration struct {
...
   DisableImageSecretPullRecheck bool // defaults to false
   PreexistingImagesPullPolicy *PreexistingImagesSecretPullPolicy
...
}

type PreexistingImageAccessPolicy string

const (
    NeverVerifyPreexistingImages PreexistingImageAccessPolicy = "NeverVerify" // the default
    IgnoreListedPreexistingImages = "IgnoreListed"
    AlwaysVerifyPreexistingImages = "AlwaysVerify"
)

type PreexistingImagesSecretPullPolicy struct {
    Type PreexistingImageAccessPolicy
    IgnoreList []string
}
```

Author: This was further noodled about during a meeting with sig-node yesterday, I'll work in the sketch we came up with.

This feature will not change the behavior of pod with `ImagePullPolicy` `Always`.

Enforcing periodical repull might be important in order to, for example, be able
to recheck image license entitlement, but is a non-goal for this enhancement.

## Proposal

The kubelet will track container images and the list of authentication information
that lead to their successful pulls. This data will be persisted across reboots
of the host and restarts of the kubelet.

The persistent authentication data storage will be done using files kept in the
kubelet directory on the node. The content of these files will be structured and
versioned using standard config file API versioning techniques.

The kubelet will ensure any image requiring credential verification is always pulled if authentication information
from an image pull is not yet present, thus enforcing authentication / re-authentication.

Two new kubelet configuration options will be added, as well as a feature
gate to gate their use:
- `imagePullCredentialsVerificationPolicy`
    - An enum of 4 possible values:
      - `NeverVerify`

>         - anyone on a node can use any image present on the node

Reviewer 2: Should we mention that this is the behavior before?

Author: added a mention of that 👍

This feature will not change the behavior of pod with `ImagePullPolicy` `Always`.

> Enforcing periodical repull might be important in order to, for example, be able

Reviewer 3: nit: periodic*

to recheck image license entitlement, but is a non-goal for this enhancement.

## Proposal

The kubelet will track container images and the list of authentication information
that lead to their successful pulls. This data will be persisted across reboots
of the host and restarts of the kubelet.

The persistent authentication data storage will be done using files kept in the
kubelet directory on the node. The content of these files will be structured and
versioned using standard config file API versioning techniques.

The kubelet will ensure any image requiring credential verification is always pulled if authentication information
from an image pull is not yet present, thus enforcing authentication / re-authentication.

Two new kubelet configuration options will be added, as well as a feature
gate to gate their use:
- `imagePullCredentialsVerificationPolicy`
    - An enum of 4 possible values:
      - `NeverVerify`
        - **original behavior before this feature**
        - anyone on a node can use any image present on the node

>       - `NeverVerifyNodePreloadedImages`

Reviewer: no strong opinion, but I'm not sure what the `Node` qualifier is adding... thoughts on shortening this to:

```suggestion
      - `NeverVerifyPreloadedImages`
```

Author: Looks like I even used the shorter version later in the text. I'll use it everywhere :+1:

        - **default value**

>         - images that were pulled to the node by something else than the kubelet can be used without reverifying pull credentials

Reviewer 2: The statement here is different with below caching policy.

In line 270-271, the Preloaded images means: image pulled without a secret by kubelet + image pulled by other processes.

Here, we only mentioned  image pulled by other processes.

Reviewer 2: My statement may be no accurate. 

- Does `image pulled without a secret by kubelet` equal to public images? public images do not need verify, which may be a common sense.

Author: That's a good point, I think we changed this a little bit. I'll change the policy definitions around line ~270 to only mention things pulled outside the kubelet process.

      - `NeverVerifyAllowlistedImages`
        - like `NeverVerifyPreloadedImages` but only node images from `preloadedImagesVerificationAllowlist` don't require reverification
      - `AlwaysVerify`

>         - all images pulled with credentials require credential reverification

Reviewer: preloaded images / images without pull records would require verification as well ... I think this can simplify to:

```suggestion
        - all images require credential reverification
```

- `preloadedImagesVerificationAllowlist`
  - a string list to be used with the `NeverVerifyAllowlistedImages` image pull credentials verification policy
  - images specified by this list will not require credential reverification if they are requested
    by a pod with `IfNotPresent` or `Never` image pull policies

>   - the list accepts wildcard suffix in a string where `*` is the wildcard character

Reviewer: do we intend to limit the wildcard to a full path segment, or are we intending to match `example.com/foo-*`?

if the former, should be something like:
```suggestion
  - the list accepts a `/*` wildcard suffix
```

>   - the list accepts wildcard suffix in a string where `*` is the wildcard character

Reviewer: I also realized some registries have leading regionalized subdomains they might want to wildcard (`foo.example.com/...`, etc). I'm on the fence about including that or not in the initial revision, but I wouldn't object to simple prefix/suffix glob. Might be worth checking main registries to see if that is widespread

Author: It looks like regionalized subdomains are quite common:
- https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview
- https://docs.oracle.com/en-us/iaas/Content/Registry/Concepts/registryprerequisites.htm
- https://cloud.google.com/artifact-registry/docs/transition/gcr-repositories#gcr-domain-support
- https://learn.microsoft.com/en-us/azure/container-registry/container-registry-geo-replication#example-use-case

I very briefly read the above documents. It seems to me that the use of regional registry is usually deliberate, and so maybe it should be in our config, too.

There was also a mention of infixed regional domains in the IBM document - see the "Private domain name" column in https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview#registry_regions_local

I'd suggest starting with suffixes for the first alpha, we can always modify the behavior to be looser.

  - granularity:
    - **finest**: image name
    - **coarsest**: registry host
  - example list values:
    - registry.host/org/image-name
    - registry.host/public-org/*
    - registry.host/*

>   - notice registry.host/org/image-name:tag would be invalid as image name is considered the most granular possible configuration

Reviewer: ```suggestion
  - Note: `registry.host/org/image-name:tag` or `registry.host/org/image-name@sha256:...` would be invalid as image name is considered the most granular possible configuration
```

1. pulled by the kubelet without auth
1. pulled by the kubelet with node-level credentials
1. images pulled by the kubelet with pod/serviceaccount-level credentials, then used again by the same pod / service account / credentials

Today, neither of the above scenarios would cause an image (or image manifest, in case the image already exists) repull in case the `ImagePullPolicy`
is set to either `IfNotPresent` or `Never`.

The goal of this enhancement is to allow multiple tenants to live on the same node without
needing to worry about tenant A's image being reused by tenant B, but at the same time to avoid unnecessary
pulls so that registries' availability does not become blocking.

The goal implies that for scenarios 1-3, the proposed solution must not put registries' availability in the critical path.

For scenario 4., the kubelet must remember credentials that were used to successfully pull an image. It must then require
a repull if there's an attempt to fetch the same image with different, currently unknown credentials. In case the
credentials that come with an image pull request were successfully used before, a repull at the registry must not be issued.

> Note that using the tag `:latest` is equivalent to using the image pull policy `Always.`

Author: This was kept here from the original KEP but after looking into the code, it seems that the `:latest` tag really only has a meaning in the defaulting logic. I'll remove the sentence.

### Kubelet Caching

The kubelet tracks a pulled image auth cache for the credentials that were successfully used to pull an image.
This also includes anonymous pulls and pulls with node-wide credentials. This cache will be persisted to disk using 
well-defined structure that follows standard config file API versioning techniques.
This is done to keep track of the authenticated images between kubelet restarts.

> The max size of the cache scales with the number of unique cache entries * the number of unique images that have not been garbage collected.

Reviewer 3: I think the GC piece could be a bit more explicit:
```
Cache entries for an image will be deleted when the image is garbage collected. Thus, the max size of the cache scales with the number of unique cache entries * the number of unique images

```

It is not expected that this would be a significant number of bytes. This claim shall be verified by actual use in Alpha and subsequent metrics in Beta.

For Alpha, the cache should be implemented solely in files and then be duplicated to
memory in Beta.

#### Caching Policies

Some images that already exist on the kubelet's node and are not tracked might be ignored
for credential reverification (and therefore ignored for caching).

The rules that define what should be ignored are set by the `imagePullCredentialsVerificationPolicy`:

> - `NeverVerify` disables the cache completely (this was the previous behavior)

Author: That's a very good point, I'll mention that the pull records should still be kept.

        kubelet->>A: ImagePullFailed
    else pull successful
        CRI->>kubelet: ID(img:tag)
        kubelet->>cache: WRITE _pulled_ (ID("img:tag"), secret A)
        kubelet->>cache: DELETE _pullling_ ("img:tag", secret A)
        kubelet->>A: continue starting pod
    end
    end
```

  2. The image is already present on the node and it's either allow-listed or a record
of its pull with the given credentials is present in the cache:

```mermaid
sequenceDiagram
    participant A as Pod A
    participant kubelet
    participant CRI
    participant cache as persistent<br/>cache

    A->>kubelet: request ("img:tag", secret A)
    kubelet->>CRI: request ID("img:tag")
    CRI->>kubelet: ID("img:tag")
    rect rgb(253, 253, 190)
    alt image present and set not to reverify
        kubelet->>kubelet: don't reverify
    else image present in cache
        kubelet->>cache: GET _pulled_ (ID("img:tag"), secret A)
        cache->>kubelet: <previous successful auth info>
    end
    end
    kubelet->>A: continue starting pod
```

##### Failure modes:

We should always fail safe. If there's an error reading from the disk, the record

> MUST be assumed as non-existent. This might cause an addition request to pull an image.

Reviewer: for clarity, does "non-existent" mean "treated as a preloaded image" or "treated as a kubelet-pulled image with unknown credential info"?

> MUST be assumed as non-existent. This might cause an addition request to pull an image.

Author: It should be the latter. I'll clarify that in the text.

The directory contains two further subdirectories: `pulling` and `pulled`. The former
serves for kubelet to store the state about images that it intends to pull. The
latter subdirectory should contain information about all the images that were
successfully pulled by the kubelet.

The figure below shows the Go structures that should appear in files in the
subdirectories described above. These structures **are not part of the REST API**
but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
They only ever appear as JSON structures in files outside the kubelet memory.

`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
directory. The filename of each such file is a SHA-256 hash of the image spec prepended
by `sha256~`.

`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
directory. The filename of each such file is a sanitized image ID provided by the
container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.

```go

> type ImagePullIntent struct {

Reviewer: is it possible for multiple parallel pull attempts of the same image spec with different credentials to happen in the kubelet? if so, how are those represented in a single file?

Reviewer: Talking through this, if intent-to-pull *only* records the image, and we handle a crash mid-pull by treating everything with an intent-to-pull record as needing reverification, we don't need *any* credential info in the intent file, and multiple parallel pulls of the same image can just ensure the intent file exists.

On kubelet start, we'd still want to look at dangling intent-to-pull records, and check if images for those exist on the node.
1. For dangling intent-to-pull images which *do* exist, create digest entries for those images indicating reverification was needed (this implies the ImagePulledRecord must be able to express "needs reverification", similar to what we'll do when we cannot read/decode an ImagePulledRecord), then clean up intent-to-pull
2. For dangling intent-to-pull entries which *don't* have an image, we could leave the intent entries in place (either permanently or for "a while" ...), which will indicate future attempts to use those exact same image specs should be reverified

Reviewer 2: If `NeverVerifyPreloadedImages`, dangling intent-to-pull entries which do exist, will become a preloaded image. How to distinguish between preloaded and dangling if we don't have the annotation?
- In the case of https://github.com/kubernetes/enhancements/pull/3547/files#diff-77a289ef32ead94beffe8193738841bf801db831a6383ed7de7342ccb0e1241c, sometimes user may trigger a manual image pull using crictl to pull manually.

If NeverVerifyAllowlistedImages(if not in the list), we can just do a reverification.

Reviewer: > dangling intent-to-pull entries which do exist, will become a preloaded image

that should not happen... a dangling intent-to-pull entry means the kubelet tried to pull and the image should be treated as a pulled-by-kubelet image with unknown credentials, which should require reverification on first use

    metav1.TypeMeta

>     ImageSpec string `json:"imageSpec"` // the filename is a hash of this value

Reviewer 3: can you explain why the filename is different here ( I assume because for the first time an image is pulled we don't have the ID yet)

    Credentials ImagePullCredentials `json:"credentials"`
}

type ImagePulledRecord struct {
    metav1.TypeMeta

>     ImageID string `json:"imageId"` // the filename is a filesystem sanitized image of this value

Reviewer: is there a clearer name to use for this?

Does this correspond to the image reference returned from CRI? If so, maybe call it image reference and document that connection?

```
	// PullImage pulls an image from the network to local storage using the supplied
	// secrets if necessary. It returns a reference (digest or ID) to the pulled image.
	PullImage(ctx context.Context, image ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error)
	// GetImageRef gets the reference (digest or ID) of the image which has already been in
	// the local storage. It returns ("", nil) if the image isn't in the local storage.
	GetImageRef(ctx context.Context, image ImageSpec) (string, error)
```

Is CRI *required* to return the digest from PullImage / GetImageRef, or does it just *happen* to? I'm wondering how file-safe we can assume the image id is in terms of sanitizing it for a filename.

Author: Image reference would likely be a better name for this field indeed.

I believe the CRI returns its own identification of the pulled image from `PullImage()`, which can either be directly the image repository digest, or the CRI internal ID of the image. I am no expert on container runtimes though so I'd rather have @haircommander or @mikebrow confirm my assumption.

Reviewer 3: yeah it's left this way historically, but ends up being an ImageRef. What further complicates the situation is containerd and cri-o do different things with the result of PullImage (containerd returns a digest, cri-o returns a repo+digest). Thus coming from cri-o I don't think we can assume it's file safe

>     Credentials ImagePullCredentials `json:"credentials"`

Reviewer: could probably unnest ImagePullCredentials fields into the top-level ImagePulledRecord, I don't think we're using ImagePullCredentials elsewhere

then, an ImagePulledRecord that only contains `apiVersion, kind, imageID` would mean "this was pulled by kubelet and has no other info about what credentials are valid"

}

// ImagePullCredentials serves two indexes
// - First index serves for searching for a specific Kubernetes secret that contained
//   the credentials used to pull an image. This assumes uninterrupted pulls if
//   credentials were rotated in a Kubernetes secret.
// - Second index serves for searching for a specific hash of credentials used to pull
//   an image. This assumes uninterrupted pulls if the same set of credentials is
//   spread across the cluster.
type ImagePullCredentials struct {

>     // CredentialHashes is an index of hashes of all the credentials that an image was pulled with

Reviewer: specify how the hash is derived? (algorithm, the content being hashed?)

    // +optional

>     CredentialHashes map[string]ImagePullCredentials `json:"credentialHashes"`

Author: Moved the hashes inside of a struct along with kube secret coordinates. I actually moved to a single value in the process. Do you think we should keep more per secret?

    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
    // secrets that were used to pull the image
    // +optional
    KubernetesSecretCoordinates map[KubernetesSecretCoordinates]ImagePullSecret `json:"kubernetesSecretCoordinates"`

    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide

>     // credentials

Reviewer: if true, is this mutually exclusive with the credentialHashes / kubernetesSecretCoordinates fields?

Author: I think so - I'll add that in the comment.

If my understanding is correct, this field would be turned to `true` if the credentials were pulled from the node's [DockerKeyring](https://github.com/kubernetes/kubernetes/blob/19e8e59d06cbdd067a9b5c141b674b63a09d0e5c/pkg/credentialprovider/plugins.go#L50).

[KEP-4412: Projected Service Account Tokens for Kubelet Image Credential Providers](https://github.com/kubernetes/enhancements/pull/4846) suggests the DockerKeyring might change in a way that the credentials would no longer be node-wide for some providers, and so that's something to keep in mind here. I assume a new field would be introduced for those cases.

    // +optional

>     NodeWideCredentials bool `json:"nodeWideCredentials"`

Reviewer: ```suggestion
    NodeWideCredentials bool `json:"omitempty,nodeWideCredentials"`
```

    // Anonymous is a flag denoting that the image was pulled without authentication.

>     // It is mutually exclusive to all the other fields of this structure.

Reviewer: ```suggestion
    // If true, it is mutually exclusive to all the other fields of this structure.
```

    // +optional

>     Anonymous bool `json:"anonymous"`

Reviewer: ```suggestion
    Anonymous bool `json:"omitempty,anonymous"`
```

Contents of `/var/lib/kubelet/image_manager/pulling/sha256~9f023ac6b143be2e542ca832efa4f162392e3f88c6e9e77b149398d19e2ad1e2`:

```json
{
  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
  "kind": "ImagePullIntent",
  "imageSpec": "docker.io/hello-world:latest",
  "credentials": {
    "anonymous": true
  }
}
```

Contents of `/var/lib/kubelet/image_manager/pulled/sha256~d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a`:

```json
{
  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
  "kind": "ImagePulledRecord",
  "imageId": "sha256:d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a",
  "credentials": {
    "anonymous": true
  }
}
```

#### Kubelet Cache Housekeeping

Upon kubelet start, the `pulling` part of the cache should be checked for dangling items.
These would be the records of images being pulled that do not appear among images
listed by container runtime at the start of the kubelet. If such a record exists, it
should be pruned.

**TODO**: how do we prevent the following race condition:
1. kubelet prepares for a pull by writing an intent record
2. kubelet issues an image pull request to the CRI
3. kubelet restarts

> 4. kubelet does not see the image so it removes the record from 1.

Author: I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.

Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

> 4. kubelet does not see the image so it removes the record from 1.

Reviewer: Ok, worth following up but doesn't seem like we can rely on the annotations for now. The [other discussion](https://github.com/kubernetes/enhancements/pull/4789#discussion_r1754951743) about how to handle dangling intent-to-pull records simplifies and resolves the race between a background pull and a kubelet crash.

> 4. kubelet does not see the image so it removes the record from 1.

Reviewer: > sometimes user may trigger a manual image pull using crictl to pull manually or to see the progress.

out of band pulls the kubelet never attempted to pull would be treated like preloaded images, and are out of scope for this enhancement

> 4. kubelet does not see the image so it removes the record from 1.

Reviewer 4: > I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.
> 
> Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

They should be stored in meta, if not it is an error.

#### Writing to the Cache

There are two different stages when the kubelet updates its cache:
1. **Before the image pull**
    - An image pull intent record is written befor the actual pull occurs.
    - This is is to mitigate a situation where the kubelet crashes during the pull and an
      image that gets pulled in the meantime later appears as if it were *preloaded*.
2. **After the pull**
    - Once we know the result of the pull operation, we either remove the image pull
      intent record (=> failed pull attempt), or we write a new record of the image
      being pulled successfully and remove the record of the pull intent.
    - Without this, if kubelet failed during an image pull, an existing record would
      already grant image access irregardless of this pull operation result for any next
      pull attempts.

The kubelet uses the following Golang interface to interact with the cache:
```go
// ImagePullCredentials wraps credentials that can be used to pull an image.
// One of the attributes must always be non-nil, and the others must be nil.
type ImagePullCredentials struct {
    KubernetesSecret *imagemanager.ImagePullSecret
    NodeWideCredentials *bool

>     Anonymous *bool

Reviewer: can collapse these to a single field named something like `AllowAllPods *bool`

Author: I started using `NodePodsAccessible` but I don't insist on it if `AllowAllPods` or any other feels more appropriate to the native speaker's ear (eye?).

}

// ImagePullManager keeps the state of images that were pulled and which are
// currently still being pulled.
// It should keep an internal state of images currently being pulled by the kubelet
// in order to determine whether to destroy a "pulling" record should an image
// pull fail.
type ImagePullManager interface {
    // RecordPullIntent records an intent to pull an image.
    // It returns a number of current pulls requesting the same image.

>     RecordPullIntent(image string) (uint23, error)

Reviewer: also make it very explicit what the image arg is (is it *exactly* what is in the pod spec?)

    // RecordImagePulled writes a record of an image being successfully pulled
    // with ImagePullCredentials.
    //
    // If `credentials` is nil, it means the image was pulled successfully, but
    // a reverification should be required upon next pull request. This can be
    // useful in cases where the kubelet crashed in the middle of the pull but
    // the image appears on the node after the kubelet restarts.

>     RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error

Reviewer: I think we *have* to pass in the same `image string` arg we pass to `RecordPullIntent` so refcounting can get cleaned up properly, *and* so the pulled record can partition these credentials by the image that was being pulled (dropping any tag and digest from the image spec).

We *also* need the canonical `imageRef` string returned from CRI.

```go
// image corresponds to the pod spec and the image passed to RecordPullIntent.
// imageRef is the canonical pulled/resolved id returned from the CRI.
RecordImagePulled(image string, imageRef string, credentials *ImagePullCredentials) error
```

    // RecordImagePullFailed should be called if an image failed to pull.

>     RecordImagePullFailed(image string) error

Reviewer: same comment here about the arg being exactly what is in the pod spec

    // ShouldReverifyCredentials evaluates the policy for image specified in
    // "image" and if the policy demands reverification, it checks the internal
    // cache to see if there's a record of pulling the image with the presented

>     // set of credentials.

Reviewer: or pulled in a way all pods are allowed to use.

    // Returns true if policy demands reverification and no record of the pull
    // was found in the cache.

>     ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool

Reviewer: maybe we want a different arg to ShouldReverify so we can make a single call instead of in a loop (and *shouldreverify* doesn't really need to explicitly opt into checking anonymous / node-wide)

>     ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool

Reviewer: doc the `image` arg corresponds to the pod spec

if we only call this when we've already resolved `image` → and existing `imageRef` on the node, should we pass that info in to this call so the implementation doesn't have to resolve it again?

>     ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool

Reviewer 2: In the past implementation, we make a *credentialprovider.DockerKeyring from the pullSecrets.

https://github.com/kubernetes/kubernetes/blob/21f54c3ca1ac87611e101f46d8b05682f51efbee/pkg/kubelet/kuberuntime/kuberuntime_image.go#L42-L82

> }

Reviewer: need a method on this interface for the GC loop to remove a credential record for an image id, I think

MUST be assumed as non-existent, meaning the image pull needs credential re-verification.
This might cause an additional request to pull an image. If it is successful,
the original failing record should be overridden to reflect the new success.

#### Cache Directory Structure

A new subdirectory is introduced to the kubelet's directory: `image_manager`.
The directory contains two further subdirectories: `pulling` and `pulled`. The former
serves for kubelet to store the state about images that it intends to pull. The
latter subdirectory should contain information about all the images that were
successfully pulled by the kubelet.

The figure below shows the Go structures that should appear in files in the
subdirectories described above. These structures **are not part of the REST API**
but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
They only ever appear as JSON structures in files outside the kubelet memory.

`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
directory. The filename of each such file is a SHA-256 hash of the image spec prepended

> by `sha256~`.

Author: Let's go with the `-` then, to prevent unwanted expansions :+1:

`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
directory. The filename of each such file is a sanitized image ID provided by the

> container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,

Author: Moved the filenames to hashes, looks like that's the safest we can do :-|

thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.

```go
// ImagePullIntent is a record of the kubelet attempting to pull an image.
type ImagePullIntent struct {
    metav1.TypeMeta

    // Image is the image spec from a Container's `image` field.
    // The filename is a SHA-256 hash of this value. This is to avoid filename-unsafe
    // characters like ':' and '/'.
    Image string `json:"image"`
}

// ImagePullRecord is a record of an image that was pulled by the kubelet.
//
// If there are no records in the `kubernetesSecrets` field and both `nodeWideCredentials`
// and `anonymous` are `false`, credentials must be re-checked the next time an
// image represented by this record is being requested.
type ImagePulledRecord struct {
    metav1.TypeMeta

    // ImageRef is a reference to the image represented by this file as received
    // from the CRI.
    // The filename is a filesystem-sanitized form of this value.

>     ImageRef string `json:"imageRef"`

Reviewer: maybe match the name the CRI uses to refer to this string (imageID, imageRef, etc)

Author: looks like `imageRef` is used there, too

>     // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes

Reviewer: nest / partition the credential info under a map keyed by the `registry/repo/image` that was pulled (can drop the tag and the digest from the image spec)

```
// map of imagespec (without tag/digest) to credentials valid for that imagespec
CredentialInfo map[string]CredentialInfo
```

This ensures that an anonymous pod can't reference a private registry or repo that *happens* to have a tagged public image in it, and discover that exists and what content it identifies.

    // secrets that were used to pull the image
    // +optional
    KubernetesSecrets []ImagePullSecret `json:"kubernetesSecretCoordinates"`

    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
    // credentials.
    // If true, it is mutually exclusive with Anonymous and KubernetesSecrets.
    // +optional
    NodeWideCredentials bool `json:"nodeWideCredentials,omitempty"`

    // Anonymous is a flag denoting that the image was pulled without authentication.
    // If true, it is mutually exclusive with KubernetesSecrets and NodeWideCredentials.
    // +optional

>     Anonymous bool `json:"anonymous,omitempty"`

Reviewer: can collapse these fields into a single `AllowAllPods bool` as well

    // content of the secret specified by the UID/Namespace/Name coordinates.
    CredentialHash `json:"credentialHash"`
}

```

**Examples:**<br>
Contents of `/var/lib/kubelet/image_manager/pulling/sha256~9f023ac6b143be2e542ca832efa4f162392e3f88c6e9e77b149398d19e2ad1e2`:

```json
{
  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
  "kind": "ImagePullIntent",
  "imageSpec": "docker.io/hello-world:latest",
  "credentials": {
    "anonymous": true
  }
}
```

Contents of `/var/lib/kubelet/image_manager/pulled/sha256~d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a`:

```json
{
  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
  "kind": "ImagePulledRecord",
  "imageId": "sha256:d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a",
  "credentials": {
    "anonymous": true
  }
}
```

#### Kubelet Cache Housekeeping

Upon kubelet start, the `pulling` part of the cache should be checked for dangling items.
These would be the records of images in the *pulling* state that now exist on the node
as listed by the CRI. If such a record exists, a `pulled` record is created in

> its place with all authentication-related fields set to their default values,

Reviewer: ```suggestion
its place with all authentication-related fields unset,
```

    //
    // `image` is the content of the pod's container `image` field.
    RecordImagePullFailed(image string)
    // MustAttemptImagePull evaluates the policy for the image specified in
    // `image` and if the policy demands verification, it checks the internal
    // cache to see if there's a record of pulling the image with the presented
    // set of credentials or if the image can be accessed by any of the node's pods.
    //
    // Returns true if the policy demands verification and no record of the pull
    // was found in the cache.
    //
    // `image` is the content of the pod's container `image` field.
    MustAttemptImagePull(image, imageRef string, credentials []imagemanager.ImagePullSecret) bool
    // PruneUnknownRecords deletes all of the cache ImagePulledRecords for each of the images
    // whose imageRef does not appear in the `imageList` iff such an record was last updated
    // _before_ the `until` timestamp.
    //
    // This method is only expected to be called by the kubelet's image garbage collector.
    // `until` is a timestamp created _before_ the `imageList` was requested from the CRI.

>     PruneUnknownRecords(imageList []string, until time.Time)

Reviewer: does the timestamp here mean we need to keep track of the time when a given imagePulledRecord was created? is the plan to do this in a field inside the record, or with a filesystem timestamp, or in memory?

Author: I originally intended to have this be an implementation detail of the cache, which would imply the files' "modify" metadata information in the alpha case.
Should we explicitly track the `LastUpdatedTime metav1.Time` timestamp instead to avoid fs timestamp being unreliable?

Reviewer: I think it's probably better to make the update time explicit in the file

The kubelet uses the following Golang interface to interact with the cache:
```go
// ImagePullManager keeps the state of images that were pulled and which are
// currently still being pulled.
// It should keep an internal state of images currently being pulled by the kubelet
// in order to determine whether to destroy a "pulling" record should an image
// pull fail.
type ImagePullManager interface {
    // RecordPullIntent records an intent to pull an image and should be called
    // before a pull of the image occurs.
    //
    // RecordPullIntent() should be called before every image pull. Each call of
    // RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
    //
    // `image` is the content of the pod's container `image` field.
    RecordPullIntent(image string) error
    // RecordImagePulled writes a record of an image being successfully pulled
    // with ImagePullCredentials.
    //
    // If `credentials` is nil, it means the image was pulled successfully, but
    // a reverification should be required upon next pull request. This can be
    // useful in cases where the kubelet crashed in the middle of the pull but

>     // the image appears on the node after the kubelet restarts.

Reviewer: I don't think we'd expect to call this to deal with dangling intent records recovering from a crash... that would happen internally on construction.

I think every call to RecordImagePulled *must* provide a non-nil `credentials *imagemanager.ImagePullCredentials`, with *either* a single image pull secret OR NodePodsAccessible=true

    //
    // `image` is the content of the pod's container `image` field.

>     RecordImagePulled(image, imageRef string, credentials *imagemanager.ImagePullCredentials) error

Reviewer: we may not need the error return if this is inactionable for the caller and it is fine for this particular caller to use the successfully pulled image anyway

    // RecordImagePullFailed should be called if an image failed to pull.
    //
    // Internally, it lowers its reference counter for the given image. If the
    // counter reaches zero, the pull intent record for the image is removed.
    //
    // `image` is the content of the pod's container `image` field.

>     RecordImagePullFailed(image string) error

Reviewer: we may not need the error return if this is inactionable for the caller and the caller already has a more important error to return (the image pull failed!)

    // ShouldReverifyCredentials evaluates the policy for image specified in
    // "image" and if the policy demands reverification, it checks the internal
    // cache to see if there's a record of pulling the image with the presented
    // set of credentials or if the image can be accessed by any of the node's pods.
    //
    // Returns true if the policy demands reverification and no record of the pull
    // was found in the cache.
    //
    // `image` is the content of the pod's container `image` field.

>     ShouldReverifyCredentials(image, imageRef string, credentials ... *imagemanager.ImagePullSecret) bool

Reviewer: there's internal verification that happens inside this method where the credentials are compared to the cache, but what this is *really* saying if it returns true is that a pull must be attempted

this might be a clearer name:

```suggestion
    MustAttemptImagePull(image, imageRef string, credentials []imagemanager.ImagePullSecret) bool
```

    // DestroyImagePulledRecord removes an ImagePulledRecord from the kubelet's
    // storage.
    // WARNING!
    //     This method is purposed for the node image GC to be able to remove an
    //     ImagePulledRecord of an image it deleted from the node.
    //     Using this method outside the GC context might cause security issues
    //     if the image represented by the `imageRef` still exists on the node at

>     //     the time this method is called.

Reviewer: If we go with the `remove(imageRef)` approach, document this should be called *after* the image is actually removed successfully.

An alternative is to pass in all known image refs (which the kubelet / gc periodically list anyway) and let the implementation clean up pulled records that aren't in that list. That would self-heal the leaked record case eventually, but may introduce races between image pulls and image listing.

```
GC thread:
  T0: token = intenttoprune() // token could be a timestamp or a generation, etc
  T10: knownImages = list all images
  T20: maybe gc stuff
  T30: pruneunknownrecords(knownImages, token) // prunes to knownImages+things recorded since intenttoprune was called (based on token)

kubelet pull thread(s)
  T15: kubelet records intent to pull foo
  T25: kubelet successfully pulls foo and records success (in a way that pruneunknownrecords can fold in)
```

I think the paired IntentToPrune / PruneUnknownRecords approach where the full list of images known to GC is passed in works better.

>     DestroyImagePulledRecord(imageRef string) error

Reviewer: I think the error is inactionable for the caller, consider dropping it?

    // characters like ':' and '/'.
    Image string `json:"image"`
}

// ImagePullRecord is a record of an image that was pulled by the kubelet.
//
// If there are no records in the `kubernetesSecrets` field and both `nodeWideCredentials`
// and `anonymous` are `false`, credentials must be re-checked the next time an
// image represented by this record is being requested.
type ImagePulledRecord struct {
    metav1.TypeMeta

    // ImageRef is a reference to the image represented by this file as received
    // from the CRI.
    // The filename is a SHA-256 hash of this value. This is to avoid filename-unsafe
    // characters like ':' and '/'.
    ImageRef string `json:"imageRef"`

    // CredentialMapping maps `image` to the set of credentials that it was
    // previously pulled with.
    // `image` in this case is the content of a pod's container `image` field that's
    // got its tag/digest removed.
    //
    // Example:
    //   Container requests the `hello-world:latest@sha256:91fb4b041da273d5a3273b6d587d62d518300a6ad268b28628f74997b93171b2` image:
    //     "credentialMapping": {
    //       "hello-world": { "nodePodsAccessible": true }
    //     }

>     CredentialMapping map[string][]ImagePullCredentials `json:"credentialMapping,omitempty"`

Reviewer: ```suggestion
    CredentialMapping map[string]ImagePullCredentials `json:"credentialMapping,omitempty"`
```

    KubernetesSecret *imagemanager.ImagePullSecret
    NodeWideCredentials *bool
    Anonymous *bool
}

// ImagePullManager keeps the state of images that were pulled and which are
// currently still being pulled.
// It should keep an internal state of images currently being pulled by the kubelet
// in order to determine whether to destroy a "pulling" record should an image
// pull fail.
type ImagePullManager interface {
    // RecordPullIntent records an intent to pull an image.
    // It returns a number of current pulls requesting the same image.
    RecordPullIntent(image string) (uint23, error)
    // RecordImagePulled writes a record of an image being successfully pulled
    // with ImagePullCredentials.
    //
    // If `credentials` is nil, it means the image was pulled successfully, but
    // a reverification should be required upon next pull request. This can be
    // useful in cases where the kubelet crashed in the middle of the pull but
    // the image appears on the node after the kubelet restarts.
    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
    // RecordImagePullFailed should be called if an image failed to pull.

>     RecordImagePullFailed(image string) error

Reviewer 3: Maybe mention the action we take in this situation in the commends (which I presume is remove the intent record?)

//
// If there are no records in the `kubernetesSecrets` field and both `nodeWideCredentials`
// and `anonymous` are `false`, credentials must be re-checked the next time an
// image represented by this record is being requested.
type ImagePulledRecord struct {
    metav1.TypeMeta

    // ImageRef is a reference to the image represented by this file as received
    // from the CRI.
    // The filename is a filesystem-sanitized form of this value.
    ImageRef string `json:"imageRef"`

    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
    // secrets that were used to pull the image
    // +optional
    KubernetesSecrets []ImagePullSecret `json:"kubernetesSecretCoordinates"`

    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
    // credentials.
    // If true, it is mutually exclusive with Anonymous and KubernetesSecrets.
    // +optional
    NodeWideCredentials bool `json:"nodeWideCredentials,omitempty"`

    // Anonymous is a flag denoting that the image was pulled without authentication.
    // If true, it is mutually exclusive with KubernetesSecrets and NodeWideCredentials.
    // +optional

>     Anonymous bool `json:"anonymous,omitempty"`

Reviewer 3: do we need to differentiate between these?

Author: this is more of a cosmetic change, the deserialization stays the same

    // ImageRef is a reference to the image represented by this file as received
    // from the CRI.
    // The filename is a SHA-256 hash of this value. This is to avoid filename-unsafe
    // characters like ':' and '/'.
    ImageRef string `json:"imageRef"`

    // CredentialMapping maps `image` to the set of credentials that it was
    // previously pulled with.
    // `image` in this case is the content of a pod's container `image` field that's
    // got its tag/digest removed.
    //
    // Example:
    //   Container requests the `hello-world:latest@sha256:91fb4b041da273d5a3273b6d587d62d518300a6ad268b28628f74997b93171b2` image:
    //     "credentialMapping": {
    //       "hello-world": { "nodePodsAccessible": true }
    //     }

>     CredentialMapping map[string][]ImagePullCredentials `json:"credentialMapping,omitempty"`

Author: should be just `CredentialMapping map[string]ImagePullCredentials`

- Feature implemented behind a feature flag - KubeletEnsureSecretPulledImages
- Initial e2e tests completed and enabled - No additional e2e identified as yet
- The kubelet cache is implemented solely via files. The implementation leverages
  interfaces in order to abstract the cache storage (in memory/via files/file-based database/...)
  and the storage format (JSON/CBOR/...). This is to allow us to later experiment with
  these various aspects to achieve optimal performance for a picked data transparency model.

#### Beta

- Kubelet cache is duplicated in memory and on disk. This is more resilient to I/O

>   operation errors when reading from disk.

Reviewer 2: IIUC,  for FG `KubeletEnsureSecretPulledImages`
- alpha: 
  - FG default false, imagePullCredentialsVerificationPolicy doesn't work. Behavior is same with `NeverVerify`.
  - When enabled, the imagePullCredentialsVerificationPolicy default is NeverVerifyPreloadedImages.
- beta: 
  - FG default true, imagePullCredentialsVerificationPolicy default to NeverVerifyPreloadedImages?
  - If you want to use the old behavior, change imagePullCredentialsVerificationPolicy=NeverVerify. Or you can just disable the FG. Either is OK.

There would be a behavior change when promoting the FG to beta.

  - Feature gate name: KubeletEnsureSecretPulledImages
  - Components depending on the feature gate: Kubelet
- [x] Other
  - Describe the mechanism: Disabling is possible by disabling the kubelet feature gate
    `KubeletEnsureSecretPulledImages`. This will turn off the feature completely.
    An alternative is to set the `imagePullCredentialsVerificationPolicy` kubelet
    configuration field to value `NeverVerify`, which will at least keep the pull
    record caching, which would make it easier to turn on the feature later.
  - Will enabling / disabling the feature require downtime of the control
    plane?

>     - No, only a restart of the kubelet

Reviewer 5: Running pods will not be impacted though right?  A drain is not needed IIUC?

Reviewer: correct, this only kicks in during the ensureimage step

###### What happens if we reenable the feature if it was previously rolled back?

Images pulled during the period the feature was disabled will not be present in the cache. Depending on the policy configured,
this could either incur redundant pulls/container creation failures OR would cause the feature to not work properly for

> images pulled during the time the feature was disabled.

Reviewer 6: When the policy is "AlwaysVerify" and registry is unavailable then the containers will not be able to create.  Clarify this more precisely?

"cause the feature to not work properly" - Clarify that the images pulled during this time will be treated.

Author: Good suggestion, see if the latest change addresses your comment :+1:
